### PR TITLE
ReadValue extension to support enums

### PR DESCRIPTION
**Pull request by staniewzki**
Signed-off-by: Michał Staniewski <m.staniewzki@gmail.com>

## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
This PR extends `ReadValueBE`/`ReadValueLE` functionality by implementing `ReadValueImpl` overload with SFINAE for enum types.

### Tests:

- [x] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: include/dali/core/byte_io.h [hunk 1]
```diff
@@ -21,7 +21,10 @@ namespace dali {
 
 namespace detail {
 
-template <int nbytes, bool is_little_endian, typename T>
+template <int nbytes,
+          bool is_little_endian,
+          typename T,
+          std::enable_if_t<std::is_integral<T>::value, bool> = true>
 void ReadValueImpl(T &value, const uint8_t* data) {
```
Comment (mzient) [inline include/dali/core/byte_io.h:28]: This usage of `enable_if` is typical for constructors/destructors. Here, you can go the usual way:
```suggestion
template <int nbytes, bool is_little_endian, typename T>
std::enable_if_t<std::is_integral<T>::value> ReadValueImpl(T &value, const uint8_t* data) {
```

**Review by jantonguirao — APPROVED** [review]

File: include/dali/core/byte_io.h [hunk 2]
```diff
@@ -34,6 +37,19 @@ void ReadValueImpl(T &value, const uint8_t* data) {
...
 }
 
+template <int nbytes,
+          bool is_little_endian,
+          typename T,
+          std::enable_if_t<std::is_enum<T>::value, bool> = true>
+void ReadValueImpl(T &value, const uint8_t* data) {
+  using U = std::underlying_type_t<T>;
+  static_assert(nbytes == sizeof(U),
+    "nbytes is expected to be the same as sizeof(std::underlying_type_t<T>)");
```
Comment (mzient) [inline include/dali/core/byte_io.h:47]: This does not need to be the case. The underlying type may be oversized for the enum and we might be deliberately reading a smaller representation.
```suggestion
  static_assert(nbytes <= sizeof(U),
    "`nbytes` should not exceed the size of the underlying type of the enum");
```

File: dali/core/byte_io_test.cc [hunk 3]
```diff
@@ -82,6 +82,28 @@ TEST(byte_io, read_value_float_input_stream) {
...
+TEST(byte_io, read_value_enum_overload) {
+  const uint8_t data_be[] = {0x00, 0x00, 0x00, 0x01};
+  const uint8_t data_le[] = {0x01, 0x00, 0x00, 0x00};
+  enum TestEnum {
+    VALUE_ZERO = 0,
+    VALUE_ONE = 1,
+  };
+  EXPECT_EQ(VALUE_ONE, ReadValueBE<TestEnum>(data_be));
+  EXPECT_EQ(VALUE_ONE, ReadValueLE<TestEnum>(data_le));
+}
```
Comment (mzient) [inline dali/core/byte_io_test.cc:94]: Make sure that sign-extension works:
```suggestion
  EXPECT_EQ(VALUE_NEGATIVE, ReadValueLE<TestEnum, 3>(data_le_negative));
}
```

File: dali/core/byte_io_test.cc [hunk 4]
```diff
@@ -82,6 +82,28 @@ TEST(byte_io, read_value_float_input_stream) {
   EXPECT_EQ(1.0f, ReadValueLE<float>(mis_le));
 }
 
+TEST(byte_io, read_value_enum_overload) {
+  const uint8_t data_be[] = {0x00, 0x00, 0x00, 0x01};
+  const uint8_t data_le[] = {0x01, 0x00, 0x00, 0x00};
```
Comment (mzient) [inline dali/core/byte_io_test.cc:87]:
```suggestion
  const uint8_t data_le[] = {0x01, 0x00, 0x00, 0x00};
  const uint8_t data_le_negative[4] = {0xD6, 0xFF, 0xFF, 0xFF };  // -42
```

**Review by mzient — APPROVED** [review]

**Pull request merged**